[ONE-Dockerizing] Scripts that can run onecc using docker containers (#9737)
Samsung/ONE: compiler/onecc-docker/README.md

# onecc-docker

> `onecc-docker` supports `one-cmd`, which is currently only supported by ubuntu 18.04, in more OS.

Reviewer: ```suggestion
`onecc-docker` supports `one-cmds`, which is currently only supported on ubuntu 18.04, in more OS.
```

Reviewer 2: Q) what is `in more OS.` ?

Author: > Q) what is in more OS. ?

More OSs can provide it.

Author: This means that `onecc-docker` supports running `onecc` in environments other than ubuntu:18.04.

Reviewer 2: How about
- _onecc-docker_ can expand `one-cmds` in more platforms than just Ubuntu 18.04/20.04.
- or
- We can use `one-cmds` in more platorms than just Ubuntu 18.04/20.04 through _onecc-docker_.

## Description

For now, `one-cmds` tools only support Ubuntu 18.04 and 20.04(not officially).

> So people in other environments can't use our tools unless they upgrade the OS (or install Ubuntu OS).

Reviewer 2: > they upgrade the OS

Actually, users may use recent Ubutuntu 22.04. Thus it is likely `they downgrade the OS`

Author: > Actually, users may use recent Ubutuntu 22.04. Thus it is likely they downgrade the OS

How about changing it like this?

Therefore, it is difficult for people in different environments to use our tools without using ubuntu18:04.

> Therefore, we developed `onecc` that runs using a docker so that users can use `one-cmds` more easily. This is `onecc-docker.`

Reviewer: ```suggestion
Therefore, we developed `onecc` that runs using a docker so that users can use `one-cmds` more easily. This is _onecc-docker_.
```

Reviewer 2: > we developed `onecc` that runs using a docker

I understand this as `onecc` is already running in docker

```suggestion
To overcome this limitation, we provide _onecc-docker_ that runs using a docker so that users can use `one-cmds` more widely.
```

> > Currently, `onecc-docker` is only supported in Linux environments.

Reviewer 2: Why is ther a `>` ?

Author: When I usually organize documents, I used BlockQuote for further explanation. I checked other documents, and it doesn't seem common. I'll fix it!

## Requirements

> - Linux OS

Reviewer 2: (optional) how about `Any Linux distribution` ?

Author: That sounds better!

- Docker

>     - `onecc-docker` requires the current `user ID` to be included in the `docker group` because it requires the docker-related commands to be executed without `sudo` privileges.

Reviewer 3: ```suggestion
    - Requires root privileges.
          - `onecc-docker` requires the current `user ID` to be included in the `docker group` because it requires the docker-related commands to be executed without `sudo` privileges.
          - See "[Post-installation steps for Linux](https://docs.docker.com/engine/install/linux-postinstall/)".
```

        ```
        sudo usermod -aG docker ${USER}
        ```

- Python 3.8

## Structure

### Summary 

`onecc-docker` creates a docker image with the latest release version of `ONE`, runs `onecc` inside the docker container after the container is driven to deliver the desired result to the user.

### To check the latest version of ONE

> - We use github api to get the latest release version of ONE. Create a docker file using the imported version.

Reviewer 2: Maybe `github api` -> `GitHub API`

Reviewer 2: I'm not sure about this line is really necessary.

Author: If you mean line 41, I'll remove it. Thank you!

Reviewer 2: No... it's line 42, explanation seems internal implementation(?)
Maybe I couldn't catch the whole story how the explanation is going.
I think this is related with using word `structure`

Reviewer 3: How about using a different name instead of `structure`?
I don't think the title and content match.

### Check Docker Image Presence and Build Image

- To prevent repetitive image generation, we verify that there is an `onecc` image of the latest `one` release version.

> - If not, build the docker image using the docker file.

Reviewer 2: `build` -> `rebuild` ?

Author: In my opinion, we build the docker image only once (unless the user intentionally deletes the image), what we repeat is considered to be the creation and deletion of the container. 
@jyoungyun  
What do you think?

Reviewer 3: > `build` -> `rebuild` ?

@seanshpark I can't catch your intention. :( Could you explain more?
The docker image is built only once there is no image with the same name. So I think `build` is more suitable for this step...

Reviewer 2: Well, this maybe because I'm not a good Docker user/developer.
What I see is a Docker image. from the explanation, when the `latest` is changed, it sounds like Docker image is built again. So from the though of `again`, it looks like `rebuild`.
Technically it looks like `build` not `rebuild` so, we can go with `build`

### Run Docker Container

> - When the above processes are successfully completed, we drive the docker container with the command the user wants. `onecc` runs on the user's current path and returns the result to the command line.

Reviewer 2: `processes` -> `steps` ?

Maybe descript steps in numbe bullets than using `###` style?

While reading here `## Structure` seems better with `### Steps`
I was expecting structure of this module or structure of how the docker works with _one-cmds_ but looks like procedure of codes are working.

Author: That must be clear. I'll reflect on it.

> ## Examples

Reviewer 3: What about separating this content into a separate file like `tutorial.md`?

### Preparations

> We prepare the Tensorflow file, configure file, and workflow for the test.

Reviewer 2: ```suggestion
Prepare source files like below;
```

```
$ tree
.
├── Dockerfile
├── inception_v3.pb
├── inception_v3.tflite
├── onecc-docker
├── onecc.template.cfg
└── onecc.workflow.json

0 directories, 6 files
```

> #### `-h` ,`--help`

Reviewer 2: maybe still WIP?

Author: Do you mean help message only for `onecc-docker`? This is under discussion with @jyoungyun .

Reviewer 3: Let's discuss after the code supporting the input arguments is shared. :)

# onecc-docker

_onecc-docker_ supports `one-cmds`, which is currently only supported by ubuntu 18.04, in more OS.

## Description

For now, `one-cmds` tools only support Ubuntu 18.04 and 20.04(not officially).
So, It is difficult for people in different environments to use our tools without using ubuntu18:04.

To overcome this limitation, we provide _onecc-docker_ that runs using a docker so that users can use `one-cmds` more widely.

Currently, `onecc-docker` is only supported in Linux environments.

## Requirements

- Any Linux distribution

- Docker

    - Requires root privileges.
            - `onecc-docker` requires the current `user ID` to be included in the `docker group` because it requires the docker-related commands to be executed without `sudo` privileges.
                  - See "[Post-installation steps for Linux](https://docs.docker.com/engine/install/linux-postinstall/)"
        
        ```
        sudo usermod -aG docker ${USER}
        ```

- Python 3.8

> 

Reviewer 3: ```suggestion
## Note

_onecc-docker_ is currently in incubation stage.

The onecc-docker debian package should be created with one-compiler debian package when ONE
compiler project builds. To this end, it is correct to configure the onecc-docker debian codes in
./infra/debian/compiler directory. However, we are currently working on the code, so we will
temporarily implement it in this location.

TODO: Merge this debian directory into ./infra/debian/compiler code.
```
